include/itkAttenuationImageFilter.hxx

```diff
@@ -0,0 +1,312 @@
+  const InputImageType * input = this->GetInput();
+  OutputImageType *      output = this->GetOutput();
+
+  ImageLinearConstIteratorWithIndex<TInputImage> it(input, regionForThread);
+  it.SetDirection(m_Direction);
+  it.GoToBegin();
+
+  unsigned int   inclusionLength;
+  InputIndexType start, end;
+
+  const float scanStepMM = input->GetSpacing()[m_Direction];
+
+  // do the work
+  while (!it.IsAtEnd())
+  {
+    inclusionLength = 0;
+    while (!it.IsAtEndOfLine())
+    {
+      // Advance until an inclusion is found
+      InputIndexType index = it.GetIndex();
+      if (ThreadedIsIncluded(index) && inclusionLength == 0)
+      {
+        // Step into inclusion
+        start = it.GetIndex();
+      }
+      else if (!ThreadedIsIncluded(index) && inclusionLength > 0)
+      {
+        // Step out of inclusion
+        end = it.GetIndex();
+
+        // Adjust for pixel padding
+        start[m_Direction] += m_PadLowerBounds;
+        end[m_Direction] -= m_PadUpperBounds;
+
+        // Estimate attenuation for each inclusion pixel
+        // with respect to the last pixel in the inclusion
+        while (start[m_Direction] < end[m_Direction])
+        {
+          // If no fixed estimation depth is set or too few pixels remain for fixed-depth estimation
+          // then take the attenuation between the given pixel and the last pixel in the inclusion
+          InputIndexType target = end;
+
+          if (m_FixedEstimationDepth != 0 && end[m_Direction] - start[m_Direction] > m_FixedEstimationDepth)
+          {
+            target[m_Direction] = start[m_Direction] + m_FixedEstimationDepth;
+          }
+
+          float estimatedAttenuation = ComputeAttenuation(target, start);
+
+          // Update the corresponding pixel in the metric image
+          if (estimatedAttenuation > 0.0 || m_ConsiderNegativeAttenuations)
+          {
+            // Assignment is thread-safe for internal operations because
+            // each thread writes only within its allotted output region.
+            output->SetPixel(start, estimatedAttenuation);
+          }
+
+          ++start[m_Direction];
+        }
+
+        inclusionLength = 0;
+      }
+
+      if (ThreadedIsIncluded(index))
+      {
+        ++inclusionLength;
+      }
+
+      ++it;
+    }
+    it.NextLine();
+  }
+};
+
+template <typename TInputImage, typename TOutputImage, typename TMaskImage>
+typename AttenuationImageFilter<TInputImage, TOutputImage, TMaskImage>::OutputPixelType
+AttenuationImageFilter<TInputImage, TOutputImage, TMaskImage>::ComputeAttenuation(const InputIndexType & end,
+                                                                                  const InputIndexType & start) const
+{
+  const float eps = 1e-15;
+
+  unsigned int numComponents = this->GetInput()->GetNumberOfComponentsPerPixel();
+  float        frequencyDelta = m_SamplingFrequencyMHz / (2 * (numComponents + 1)); // TODO verify +1 is correct
```

Comment:
@dzenanz I added the `+ 1` here to accommodate the fact that we drop the DC channel from spectra output. Could you please compare with `AttenuationEstimator` and verify that this is correct?

Reply:
Ping @thewtex

Reply:
As I investigate this I'm becoming more convinced that `AttenuationEstimator` is correct and the `+1` here is incorrect. I had thought that we were dropping a frequency _bin_ in discarding the DC component, but based on comments in [`itkSpectraImageFilter.h`](https://github.com/KitwareMedical/ITKUltrasound/blob/master/include/itkSpectra1DImageFilter.h#L48) and [`itkSpectraImageFilter.hxx`](https://github.com/KitwareMedical/ITKUltrasound/blob/master/include/itkSpectra1DImageFilter.hxx#L58) it seems more likely that the first of 31 bins that we have after dropping the DC component represents content in the range (0, 0 + f_delta]. In that case no bins are dropped in generating the spectra image and the frequency delta is indeed `f_nyquist / 31`.

Reply:
I think he is swamped by GitHub notifications, and if you want to catch his attention you need to send an email or talk to him in person 😄

Reply:
His calendar appears to be completely full today 😅 

I'm sufficiently convinced that the original code was correct. I will make the change and update baseline images accordingly.

Reply:
@dzenanz Could you please re-review specifically the attenuation calculation method? Would like to verify that the comments I added are clear and correct.